### PR TITLE
Refactor type detection in DTO test

### LINKED ISSUE #11160: Compilation errors with DTO, OAuth and ManyToMany with User
##### **Overview of the issue**

Compilation errors below:

```
[ERROR] /Users/theo/Documents/perso/jhipster-many/src/test/java/com/mycompany/myapp/service/mapper/CountryMapperTest.java:[20,45] incompatible types: java.lang.String cannot be converted to java.lang.Long
[ERROR] /Users/theo/Documents/perso/jhipster-many/src/test/java/com/mycompany/myapp/service/dto/CountryDTOTest.java:[14,53] incompatible types: java.lang.String cannot be converted to java.lang.Long
[ERROR] /Users/theo/Documents/perso/jhipster-many/src/test/java/com/mycompany/myapp/service/dto/CountryDTOTest.java:[19,53] incompatible types: java.lang.String cannot be converted to java.lang.Long
```

##### **Motivation for or Use Case**

Project should compile.

##### **Reproduce the error**

Generate the app using the JDL below:

<details>
<summary>JDL file</summary>
<pre>
application {
  config {
    authenticationType oauth2
  }
  entities *
}

entity Country {
    name String
}

relationship ManyToMany {
    Country{User(login)} to User{Country(name)}
}

dto Country with mapstruct</pre>
</details>

##### **JHipster Version(s)**

```
jhipster@0.0.1-SNAPSHOT /Users/theo/Documents/perso/jhipster-many/mono
└── generator-jhipster@6.6.0

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "jhiPrefix": "jhi",
    "languages": ["en", "fr"],
    "messageBroker": false,
    "nativeLanguage": "en",
    "packageName": "com.mycompany.myapp",
    "packageFolder": "com/mycompany/myapp",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "serviceDiscoveryType": false,
    "skipClient": false,
    "skipServer": false,
    "testFrameworks": [],
    "websocket": false,
    "authenticationType": "oauth2",
    "jhipsterVersion": "6.6.0",
    "baseName": "jhipster",
    "buildTool": "maven",
    "skipUserManagement": true,
    "clientPackageManager": "npm",
    "applicationType": "monolith",
    "cacheProvider": "ehcache",
    "clientFramework": "angularX",
    "clientTheme": "none",
    "clientThemeVariant": "",
    "serverPort": "8080",
    "useSass": true,
    "embeddableLaunchScript": false,
    "creationTimestamp": 1579723742179,
    "entitySuffix": "",
    "dtoSuffix": "DTO",
    "otherModules": [],
    "blueprints": []
  },
  "entities": ["Country"]
}

</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Country {
  name String
}
relationship ManyToMany {
  Country{user(login)} to User
}

dto Country with mapstruct
service Country with serviceClass

</pre>
</details>

##### **Environment and Tools**

openjdk version "12.0.2" 2019-07-16
OpenJDK Runtime Environment AdoptOpenJDK (build 12.0.2+10)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.15.1, JRE 12 Mac OS X amd64-64-Bit Compressed References 20190915_201 (JIT enabled, AOT enabled)
OpenJ9   - 0f66c6431
OMR      - ec782f26
JCL      - 9984ea0dd9 based on jdk-12.0.2+10)

git version 2.25.0

node: v10.18.1

npm: 6.13.4

yarn: 1.21.1

Docker version 19.03.5, build 633a0ea

docker-compose version 1.24.1, build 4667896b

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by Falydoor**
-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

Fix #11160

### CODE REVIEW

File: generators/generator-base-private.js [hunk 1]
```diff
@@ -1463,6 +1463,34 @@ module.exports = class extends Generator {
...
+                // Special case with a OneToOne relationship with User and @MapsId when using OAuth
+                if (databaseType === 'sql') {
+                    value = 'UUID.randomUUID().toString()';
+                }
+                break;
+            case 'UUID':
+                value = 'UUID.randomUUID()';
+                break;
+            default:
+                value = `${defaultValue}L`;
```
Comment (SudharakaP) [inline generators/generator-base-private.js]: Is the letter `L` at the end a typo?

Comment (MathieuAA) [reply generators/generator-base-private.js]: Don't think so, `L` may mean the value is a Long.

File: generators/generator-base-private.js [hunk 2]
```diff
@@ -1463,6 +1463,34 @@ module.exports = class extends Generator {
...
     }
 
+    /**
+     * Returns the primary key value based on authentication type, DB, given association and default value
+     *
+     * @param {string} authenticationType - the auth type
+     * @param {string} databaseType - the database type
+     * @param {any} relationships - relationships
+     * @param {string} defaultValue - default value
+     */
```
Comment (MathieuAA) [inline generators/generator-base-private.js]: Can you add the `@return` tag please

File: generators/generator-base-private.js [hunk 3]
```diff
@@ -1463,6 +1463,34 @@ module.exports = class extends Generator {
...
+    getPkValueBasedOnDBAndAssociation(authenticationType, databaseType, relationships, defaultValue) {
```
Comment (MathieuAA) [inline generators/generator-base-private.js]: An idea:
  - the function has way too many arguments, and most of them (`authenticationType` and `relationships`) are only used to get the PK type
  - the PK type can be passed to this function as arg
  - the result `getBlablabla(primaryKeyType, databaseType, defaultValue)` is better, but still: the name doesn't fit
  - `getPrimaryKeyValue(...)` is an improvement over the previous version
  - you can further refactor this, but as it's only used once or twice... there may not be any need for breaking the switch-case into something like getPrimaryKeyValueForStringKey, or ForUUIDKey, etc. (you may have to keep this in mind if you plan to refactor more than one file though)

By the way, **where are the tests?**

**Comment by MathieuAA** [comment]
@Falydoor Hello! @qmonmert has a PR for this issue too. Do you think you have some time to finish your PR? The difference between the two PRs is that yours adds a nice refactoring.
If you don't have time to continue, I propose to merge Quentin's PR first, and add your refactoring later.
Are you okay? We can wait a bit, no issue.
Thank you guys for the PRs!

**Comment by qmonmert** [comment]
@MathieuAA @Falydoor  ok it's up to you :)

**Comment by Falydoor** [comment]
Hello @MathieuAA,

I was on vacation in London so did not really open my laptop!

I'm gonna add tests and do the changes based on your comment.

**Comment by MathieuAA** [comment]
Thanks!

**Comment by Falydoor** [comment]
@MathieuAA I enabled a more complete entity test for oauth2 that covers the initial issue. I also fixed few issues that were found because of the test.

Let me know if more is needed!

File: .github/workflows/angular.yml [hunk 4]
```diff
@@ -67,7 +67,7 @@ jobs:
                       war: 0
                       protractor: 1
                     - app-type: ngx-mariadb-oauth2-infinispan
-                      entity: sql
+                      entity: sqlfull
```
Comment (MathieuAA) [inline .github/workflows/angular.yml]: why is it in your PR?

File: generators/generator-base.js [hunk 5]
```diff
@@ -1426,6 +1426,12 @@ module.exports = class extends PrivateBase {
             );
 
             limit = 61;
+        } else if (prodDatabaseType === 'mariadb' && constraintName.length >= 61 && !this.skipCheckLengthOfIdentifier) {
+            this.warning(
+                `The generated constraint name "${constraintName}" is too long for MariaDB (which has a 64 characters limit). It will be truncated!`
```
Comment (MathieuAA) [inline generators/generator-base.js]: Small typo: `64 character limit`

**Comment by MathieuAA** [comment]
Everything looks good (except the small typo), thanks!

**Comment by Falydoor** [comment]
Bounty claimed https://opencollective.com/generator-jhipster/expenses/13409

**Pull request merged**